[FSDP] Handle the state_dict on CPU cases

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85761
* #85740
* __->__ #85640

state_dict may not be on GPUs. We need to move it to the compute_device in order to gather the ShardedTensor.

Differential Revision: [D39681730](https://our.internmc.facebook.com/intern/diff/D39681730/)

fegin committed:
[FSDP] Handle the state_dict on CPU cases

state_dict may not be on GPUs. We need to move it to the compute_device in order to gather the ShardedTensor.

Differential Revision: [D39681730](https://our.internmc.facebook.com/intern/diff/D39681730/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -482,15 +484,28 @@ def _dist_train(self, wrap_fsdp: bool, state_dict_type: str = ""):
-        fsdp_params = self._dist_train(wrap_fsdp=True, state_dict_type=state_dict_type)
+    @parametrize("move_to_cpu", [True, False])
+    def test_state_dict_save_load_flow(self, state_dict_type, move_to_cpu):
+        fsdp_params = self._dist_train(
```
nit: instead of parametrize, use 

```
for move_to_cpu in [True, False]:
    with self.subTest(move_to_cpu=move_to_cpu):
        # test code
```

as this reuses the processes instead of re-spawning them

rohan-varma reviewed (approved):
LGTM

fegin committed:
Update on "[FSDP] Handle the state_dict on CPU cases"


state_dict may not be on GPUs. We need to move it to the compute_device in order to gather the ShardedTensor.

Differential Revision: [D39681730](https://our.internmc.facebook.com/intern/diff/D39681730/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Handle the state_dict on CPU cases"


state_dict may not be on GPUs. We need to move it to the compute_device in order to gather the ShardedTensor.

Differential Revision: [D39681730](https://our.internmc.facebook.com/intern/diff/D39681730/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -482,6 +484,16 @@ def _dist_train(self, wrap_fsdp: bool, state_dict_type: str = ""):
             _zero_model(blank_model)
             state_dict = self._state_dict(model, state_dict_type)
+            if move_to_cpu:
+                for key in list(state_dict.keys()):
```
Curious: why `list(state_dict.keys())` and not just `state_dict.keys()`?

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -482,6 +484,16 @@ def _dist_train(self, wrap_fsdp: bool, state_dict_type: str = ""):
+            if move_to_cpu:
+                for key in list(state_dict.keys()):
+                    tensor = state_dict[key]
+                    if isinstance(tensor, torch.Tensor):
```
`ShardedTensor` returns `False` for `isinstance(..., torch.Tensor)`?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2533,15 +2533,20 @@ def _sharded_pre_load_state_dict_hook(
+                if num_padding > 0:
+                    local_tensor = F.pad(local_tensor, [0, num_padding])
+            else:
+                local_tensor = torch.zeros(chunk_size, dtype=param.dtype).cuda()
```
nit: May save an H2D copy
```suggestion
                local_tensor = torch.zeros(chunk_size, dtype=param.dtype, device=self.compute_device)
```
or `device=flat_param.device` (I do not think it should matter)

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -482,6 +484,16 @@ def _dist_train(self, wrap_fsdp: bool, state_dict_type: str = ""):
+                        state_dict[key] = tensor.cpu()
+                    else:
+                        shards = tensor.local_shards()
+                        if shards:
```
Do we want an assertion here like `assert len(shards) == 1` to remind us that this test needs to change if we support multiple local shards?

awgu reviewed (approved):
LGTM!

Does this conflict with the `TensorFlattener` PR?
